File: engine/src/theseus_engine/controller.py
"""NavController: wires perception, world model, planners, steering,
guidance and the FSM into a per-tick loop.

On iOS this file becomes the actor pipeline (perception actor -> world
model actor -> planner actor -> guidance/steering at cue rate) described
in docs/ARCHITECTURE.md; the logic and the order of operations per tick
are meant to survive that translation:

  advance world -> sense -> notify planner of changes -> validate current
  path -> (replan incrementally if needed, via the FSM) -> compute cue ->
  steer -> move -> emit trace frame.

Mapping uses optimistic planning (unknown is traversable) with periodic
full A* — you are allowed to *plan into* space you haven't seen while
building the map. Guidance is pessimistic (unknown is a wall) and leans
on D* Lite so mid-route world changes cost milliseconds, not a rebuild.
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, replace

from . import astar, frontier
from .dstar_lite import DStarLite
from .fsm import Event, State, StateMachine
from .geometry import Cell, Vec, bearing, dist, wrap_angle
from .grid import OCCUPIED, PlanParams
from .guidance import ARRIVE, OFF_ROUTE, GuidanceFollower
from .sim import Simulator
from .steering import VFHSteering
from .trace import TraceWriter


@dataclass
class ControllerConfig:
    map_replan_every: int = 40      # ticks between scheduled mapping replans
    clearance_every: int = 5        # ticks between clearance refreshes
    wp_arrive: float = 0.5          # m: "reached this scan waypoint"
    blocked_wait: int = 15          # ticks to wait in BLOCKED before RETRY
    blocked_steer_limit: int = 8    # consecutive boxed-in ticks -> force replan
    guidance_lookahead: float = 0.9
    pivot_thresh: float = 0.45      # rad: heading error above this -> turn in place


class NavController:
    def __init__(self, sim: Simulator, params: PlanParams,
                 trace: TraceWriter | None = None,
                 config: ControllerConfig | None = None):
        self.sim = sim
        self.params = params
        self.opt_params = replace(params, unknown_ok=True)
        self.cfg = config or ControllerConfig()
        self.fsm = StateMachine()
        self.trace = trace
        self.steer_guide = VFHSteering(params)
        self.steer_map = VFHSteering(self.opt_params)
        self.replans = 0
        self.frames = 0

    # ---- helpers ---------------------------------------------------------

    def _pos(self) -> Vec:
        return (self.sim.pose[0], self.sim.pose[1])

    def _cell(self) -> Cell:
        return self.sim.est.world_to_cell(self._pos())

    def _tick_world(self) -> list[Cell]:
        self.sim.advance()
        changed = self.sim.sense()
        if self.sim.tick % self.cfg.clearance_every == 0:
            self.sim.est.refresh_clearance()
        return changed

    def _emit(self, changed, cue=None, steer=None, events=(),
              path=None, smoothed=None, replan_ms=None) -> None:
        if self.trace is None:
            return
        x, y, th = self.sim.pose
        frame = {
            "t": round(self.sim.tick * self.sim.dt, 3),
            "pose": [x, y, th],
            "state": self.fsm.state.value,
            "occ": [[c[0], c[1], self.sim.est.state(c)] for c in changed],
            "movers": self.sim.mover_positions(),
        }
        if cue is not None:
            frame["cue"] = {"kind": cue.kind, "dist": cue.distance,
                            "deg": cue.angle_deg, "ct": cue.cross_track,
                            "corridor": cue.corridor,
                            "target": [cue.target[0], cue.target[1]]}
        if steer is not None:
            frame["steer"] = {"h": steer.heading, "v": steer.speed,
                              "free": steer.free_dist,
                              "blocked": steer.blocked}
        if events:
            frame["events"] = list(events)
        if path is not None:
            frame["path"] = [[c[0], c[1]] for c in path]
        if smoothed is not None:
            frame["smoothed"] = [[p[0], p[1]] for p in smoothed]
        if replan_ms is not None:
            frame["replan_ms"] = replan_ms
        self.trace.frame(**frame)
        self.frames += 1

    def _move(self, steer) -> None:
        """Steering validated free space along steer.heading — not along
        directions passed through while rotating. So: large heading error
        means pivot in place first, advance only once roughly aligned."""
        if steer.blocked:
            self.sim.step_motion(self.sim.pose[2], 0.0)
            return
        err = abs(wrap_angle(steer.heading - self.sim.pose[2]))
        speed = 0.0 if err > self.cfg.pivot_thresh else steer.speed
        self.sim.step_motion(steer.heading, speed)

    @staticmethod
    def _remaining(cells: list[Cell], ptr: int, here: Cell) -> int:
        """Advance a monotone pointer to the path cell nearest the agent
        (looking a bounded window ahead), so validity checks only consider
        the part of the path still to be walked."""
        best_i, best_d = ptr, None
        for i in range(ptr, min(ptr + 12, len(cells))):
            d = abs(cells[i][0] - here[0]) + abs(cells[i][1] - here[1])
            if best_d is None or d < best_d:
                best_i, best_d = i, d
        return best_i

    # ---- mapping phase -----------------------------------------------------

    def run_mapping(self, scan_pts: list[Vec], laps: int = 2,
                    per_wp: int = 350, done_check=None) -> bool:
        """Walk the scan waypoints (optimistically planned) to build the
        map, like a user sweeping their phone around a room. Stops early
        if done_check() says the map is good enough."""
        est = self.sim.est
        self.fsm.step(Event.SCAN_STARTED)
        for _lap in range(laps):
            for wp in scan_pts:
                self._goto_optimistic(tuple(wp), per_wp)
            est.refresh_clearance(force=True)
            if done_check is not None and done_check():
                break
        self.fsm.step(Event.MAP_READY)
        return True

    def _goto_optimistic(self, wp: Vec, max_ticks: int) -> bool:
        est = self.sim.est
        follower = None
        cells: list[Cell] | None = None
        ptr = 0
        age = 10 ** 9
        for _ in range(max_ticks):
            changed = self._tick_world()
            if dist(self._pos(), wp) < self.cfg.wp_arrive:
                self._emit(changed)
                return True
            age += 1
            need = (cells is None or age >= self.cfg.map_replan_every
                    or not astar.path_valid(est, cells[ptr:], self.opt_params))
            events = []
            replan_ms = None
            path_field = smoothed_field = None
            if need:
                est.refresh_clearance()
                t0 = time.perf_counter()
                res = astar.plan(est, self._cell(), est.world_to_cell(wp),
                                 self.opt_params)
                replan_ms = (time.perf_counter() - t0) * 1000.0
                self.replans += 1
                age = 0
                if res is None:
                    self._emit(changed, events=["WP_UNREACHABLE"],
                               replan_ms=replan_ms)
                    return False  # try again next lap
                cells, ptr = res.cells, 0
                smoothed = astar.smooth(est, cells, self.opt_params)
                follower = GuidanceFollower(smoothed,
                                            lookahead=self.cfg.guidance_lookahead)
                events.append("PLANNED")
                path_field, smoothed_field = cells, smoothed
            ptr = self._remaining(cells, ptr, self._cell())
            cue = follower.cue(est, self.sim.pose)
            if cue.kind == ARRIVE:
                self._emit(changed, cue=cue, events=events)
                return True
            steer = self.steer_map.decide(est, self.sim.pose,
                                          bearing(self._pos(), cue.target))
            self._move(steer)
            if cue.kind == OFF_ROUTE:
                age = 10 ** 9  # force replan next tick from wherever we are
            self._emit(changed, cue=cue, steer=steer, events=events,
                       path=path_field, smoothed=smoothed_field,
                       replan_ms=replan_ms)
        return False

    # ---- explore phase (virtual-agent auto-mapping) ------------------------

    def run_explore(self, max_targets: int = 60, per_target: int = 400,
                    min_cluster: int = 4) -> dict:
        """Frontier-driven auto-mapping: walk to the nearest frontier
        cluster until none remain. Target ranking is pessimistic (only
        chase frontiers reachable through known space); motion reuses the
        optimistic mapping primitive — mapping-mode semantics, same as a
        user sweeping their phone, just self-directed."""
        est = self.sim.est
        self.fsm.step(Event.SCAN_STARTED)
        changed = self._tick_world()          # open our eyes before asking
        est.refresh_clearance(force=True)     # "is there anything to see?"
        self._emit(changed, events=["EXPLORE_STARTED"])
        visited = 0
        failures = 0
        while visited < max_targets:
            est.refresh_clearance(force=True)
            tgt = frontier.select_target(est, self._cell(), self.params,
                                         min_cluster=min_cluster,
                                         min_dist_m=self.cfg.wp_arrive + 0.25)
            if tgt is None:
                break
            visited += 1
            if self._goto_optimistic(est.cell_center(tgt.cell), per_target):
                failures = 0
            else:
                failures += 1
                if failures >= 3:
                    break  # the remaining clusters keep defeating us; stop
        self.fsm.step(Event.MAP_READY)
        return {"targets": visited, "failures": failures}

    # ---- walk mode (free roam, no destination) ------------------------------

    def run_walk(self, ticks: int) -> dict:
        """Walk mode: no destination — keep moving through free space,
        reacting to whatever the sensor sees. This is the original
        "longest unobstructed forward vector" idea made stable: VFH picks
        the most open sector and hysteresis keeps the choice from
        flickering. Uses mapping-mode optimism (unknown space ahead is
        walkable until the continuously-running sensor objects) because
        reactive avoidance is the sensor's job here, not the map's."""
        self.fsm.step(Event.WALK_TOGGLED)
        traveled = 0.0
        blocked_ticks = 0
        for _ in range(ticks):
            changed = self._tick_world()
            before = self._pos()
            steer = self.steer_map.decide(self.sim.est, self.sim.pose, None)
            # Two safety gates before advancing, both walk-mode-specific:
            # (1) don't outrun the sensor — if the ground along the chosen
            #     heading wasn't SEEN recently, pivot first so the FOV
            #     sweeps it (a mover may have wandered into stale space);
            # (2) swept-body check against LIVE cell states — VFH's thin
            #     ray can miss a body-width graze, and the clearance field
            #     it leans on is cached (refreshed every few ticks), which
            #     is exactly long enough for a walking person to shift one
            #     body width. Check actual occupied cells, not the cache.
            if not steer.blocked and (not self._front_fresh(steer.heading)
                                      or not self._swept_clear(steer.heading)):
                steer = replace(steer, speed=0.0, reason="hold_safe")
            self._move(steer)
            traveled += dist(before, self._pos())
            blocked_ticks += int(steer.blocked)
            self._emit(changed, steer=steer)
        self.fsm.step(Event.STOP)
        return {"traveled_m": round(traveled, 2),
                "blocked_ticks": blocked_ticks}

    def _swept_clear(self, heading: float, advance_m: float = 0.3,
                     pad: float = -0.02) -> bool:
        """True if the body-swept corridor along `heading` has no
        OCCUPIED cell INSIDE the legal clearance minimum, judged on
        current cell states (which update every sense) instead of the
        cached clearance field (which does not). The gate radius sits
        just UNDER the body radius on purpose: skimming a wall at the
        cost model's legal 0.28 m is allowed; a person's cells intruding
        deep into the corridor — the thing the stale cache misses — is
        not. A wider radius here re-vetoes legal passages and the walker
        stalls (measured: 264/350 ticks held at +0.03 pad)."""
        est = self.sim.est
        x, y, _ = self.sim.pose
        r = self.params.radius + pad
        span = int(r / est.cell_size) + 1
        ca, sa = math.cos(heading), math.sin(heading)
        step = est.cell_size
        d = step
        while d <= advance_m:
            px, py = x + ca * d, y + sa * d
            pc = est.world_to_cell((px, py))
            for dy in range(-span, span + 1):
                for dx in range(-span, span + 1):
                    c = (pc[0] + dx, pc[1] + dy)
                    if (est.in_bounds(c) and est.state(c) == OCCUPIED
                            and dist(est.cell_center(c), (px, py)) <= r):
                        return False
            d += step
        return True

    def _front_fresh(self, heading: float, dist_m: float = 0.5,
                     max_age: int = 8) -> bool:
        """True if every cell within dist_m along `heading` was observed
        within the last max_age ticks."""
        est = self.sim.est
        x, y, _ = self.sim.pose
        ca, sa = math.cos(heading), math.sin(heading)
        steps = max(1, int(dist_m / est.cell_size))
        for i in range(1, steps + 1):
            d = i * est.cell_size
            c = est.world_to_cell((x + ca * d, y + sa * d))
            if not est.in_bounds(c):
                return True          # the boundary is a wall, not a mover
            if self.sim.tick - est.last_seen[est.idx(c)] > max_age:
                return False
        return True

    # ---- guidance phase ("Ariadne mode") -----------------------------------

    def run_guidance(self, goal: Vec, max_ticks: int = 2500) -> bool:
        """Guide toward `goal` with incremental replanning. Returns True
        on arrival."""
        est = self.sim.est
        est.refresh_clearance(force=True)
        goal_cell = est.world_to_cell(goal)
        self.fsm.step(Event.GOAL_SET)

        ds = DStarLite(est, self._cell(), goal_cell, self.params)
        follower: GuidanceFollower | None = None
        cells: list[Cell] | None = None
        ptr = 0
        boxed = 0
        wait = 0

        def replan(events: list[str]) -> tuple[list[Cell] | None, float]:
            nonlocal follower, ptr
            est.refresh_clearance()
            t0 = time.perf_counter()
            res = ds.plan()
            ms = (time.perf_counter() - t0) * 1000.0
            self.replans += 1
            if res is None:
                if self.fsm.state != State.BLOCKED:
                    self.fsm.step(Event.PLAN_FAILED)
                events.append("PLAN_FAILED")
                return None, ms
            smoothed = astar.smooth(est, res.cells, self.params)
            follower = GuidanceFollower(smoothed,
                                        lookahead=self.cfg.guidance_lookahead)
            ptr = 0
            self.fsm.step(Event.PLAN_READY)
            events.append("REROUTED")
            return res.cells, ms

        events: list[str] = []
        cells, ms = replan(events)
        self._emit([], events=events, path=cells,
                   smoothed=follower.path if follower else None, replan_ms=ms)

        for _ in range(max_ticks):
            changed = self._tick_world()
            if changed:
                ds.notify_changed(changed)
            here = self._cell()
            if here != ds.start:
                ds.update_start(here)

            events = []
            replan_ms = None
            path_field = smoothed_field = None

            if self.fsm.state == State.BLOCKED:
                wait += 1
                if wait >= self.cfg.blocked_wait:
                    wait = 0
                    self.fsm.step(Event.RETRY)   # BLOCKED -> PLANNING
                    cells, replan_ms = replan(events)
                    path_field = cells
                    smoothed_field = follower.path if cells else None
                self._emit(changed, events=events, path=path_field,
                           smoothed=smoothed_field, replan_ms=replan_ms)
                continue

            ptr = self._remaining(cells, ptr, here)
            if not astar.path_valid(est, cells[ptr:], self.params):
                self.fsm.step(Event.ROUTE_BLOCKED)   # GUIDING -> PLANNING
                cells, replan_ms = replan(events)
                path_field = cells
                smoothed_field = follower.path if cells else None
                if cells is None:
                    self._emit(changed, events=events, replan_ms=replan_ms)
                    continue

            cue = follower.cue(est, self.sim.pose)
            if cue.kind == ARRIVE:
                self.fsm.step(Event.ARRIVED_EVT)
                self._emit(changed, cue=cue, events=events + ["ARRIVED"],
                           path=path_field, smoothed=smoothed_field,
                           replan_ms=replan_ms)
                return True
            if cue.kind == OFF_ROUTE:
                self.fsm.step(Event.OFF_ROUTE)      # GUIDING -> PLANNING
                cells, replan_ms = replan(events)
                path_field = cells
                smoothed_field = follower.path if cells else None
                if cells is None:
                    self._emit(changed, events=events, replan_ms=replan_ms)
                    continue
                cue = follower.cue(est, self.sim.pose)

            steer = self.steer_guide.decide(est, self.sim.pose,
                                            bearing(self._pos(), cue.target))
            if steer.blocked:
                boxed += 1
                if boxed >= self.cfg.blocked_steer_limit:
                    boxed = 0
                    self.fsm.step(Event.ROUTE_BLOCKED)
                    cells, replan_ms = replan(events)
                    path_field = cells
                    smoothed_field = follower.path if cells else None
            else:
                boxed = 0
            self._move(steer)
            self._emit(changed, cue=cue, steer=steer, events=events,
                       path=path_field, smoothed=smoothed_field,
                       replan_ms=replan_ms)
        return False

File: apps/navcore/Sources/NavCore/Controller.swift
// NavController — port of engine controller.py: wires perception, world
// model, planners, steering, guidance and the FSM into a per-tick loop.
// On iOS this becomes the actor pipeline; the order of operations per
// tick is meant to survive that translation:
//   advance world -> sense -> notify planner -> validate path ->
//   (replan via FSM) -> cue -> steer -> move -> emit trace frame.
//
// replan_ms is volatile (excluded from golden comparison), so this port
// simply omits it — behavior parity is unaffected.

#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#elseif os(Windows)
import CRT
#endif

public struct ControllerConfig: Sendable {
    public var mapReplanEvery = 40    // ticks between scheduled map replans
    public var clearanceEvery = 5     // ticks between clearance refreshes
    public var wpArrive = 0.5         // m: "reached this scan waypoint"
    public var blockedWait = 15       // ticks in BLOCKED before RETRY
    public var blockedSteerLimit = 8  // consecutive boxed-in -> force replan
    public var guidanceLookahead = 0.9
    public var pivotThresh = 0.45     // rad: above this -> turn in place

    public init() {}
}

public func pyRound3(_ x: Double) -> Double {
    let r = (x * 1e3).rounded(.toNearestOrEven) / 1e3
    return r == 0 ? 0.0 : r
}

public final class NavController {
    public let sim: Simulator
    public let params: PlanParams
    public let optParams: PlanParams
    public let cfg: ControllerConfig
    public let fsm = StateMachine()
    public let trace: TraceWriter?
    public let steerGuide: VFHSteering
    public let steerMap: VFHSteering
    public private(set) var replans = 0
    public private(set) var frames = 0

    public init(sim: Simulator, params: PlanParams,
                trace: TraceWriter? = nil,
                config: ControllerConfig = ControllerConfig()) {
        self.sim = sim
        self.params = params
        var opt = params
        opt.unknownOk = true
        self.optParams = opt
        self.cfg = config
        self.trace = trace
        self.steerGuide = VFHSteering(params: params)
        self.steerMap = VFHSteering(params: opt)
    }

    // ---- helpers ---------------------------------------------------------

    func pos() -> Vec {
        Vec(sim.pose.0, sim.pose.1)
    }

    func cell() -> Cell {
        sim.est.worldToCell(pos())
    }

    func tickWorld() -> [Cell] {
        sim.advance()
        let changed = sim.sense()
        if sim.tick % cfg.clearanceEvery == 0 {
            sim.est.refreshClearance()
        }
        return changed
    }

    func emit(_ changed: [Cell], cue: GuidanceCue? = nil,
              steer: SteeringDecision? = nil, events: [String] = [],
              path: [Cell]? = nil, smoothed: [Vec]? = nil) {
        guard let trace else { return }
        let (x, y, th) = sim.pose
        var frame: [String: TraceValue] = [
            "t": .double(pyRound3(Double(sim.tick) * sim.dt)),
            "pose": .array([.double(x), .double(y), .double(th)]),
            "state": .string(fsm.state.rawValue),
            "occ": .array(changed.map { c in
                .array([.int(Int(c.x)), .int(Int(c.y)),
                        .int(sim.est.state(c))])
            }),
            "movers": .array(sim.moverPositions().map { m in
                .array(m.map { .double($0) })
            }),
        ]
        if let cue {
            frame["cue"] = .object([
                "kind": .string(cue.kind.rawValue),
                "dist": .double(cue.distance),
                "deg": .double(cue.angleDeg),
                "ct": .double(cue.crossTrack),
                "corridor": .double(cue.corridor),
                "target": .array([.double(cue.target.x),
                                  .double(cue.target.y)]),
            ])
        }
        if let steer {
            frame["steer"] = .object([
                "h": .double(steer.heading),
                "v": .double(steer.speed),
                "free": .double(steer.freeDist),
                "blocked": .bool(steer.blocked),
            ])
        }
        if !events.isEmpty {
            frame["events"] = .array(events.map { .string($0) })
        }
        if let path {
            frame["path"] = .array(path.map { c in
                .array([.int(Int(c.x)), .int(Int(c.y))])
            })
        }
        if let smoothed {
            frame["smoothed"] = .array(smoothed.map { p in
                .array([.double(p.x), .double(p.y)])
            })
        }
        trace.frame(frame)
        frames += 1
    }

    /// Steering validated free space along steer.heading — not along
    /// directions passed through while rotating. Large heading error
    /// means pivot in place first, advance only once roughly aligned.
    func move(_ steer: SteeringDecision) {
        if steer.blocked {
            _ = sim.stepMotion(headingCmd: sim.pose.2, speedCmd: 0.0)
            return
        }
        let err = abs(wrapAngle(steer.heading - sim.pose.2))
        let speed = err > cfg.pivotThresh ? 0.0 : steer.speed
        _ = sim.stepMotion(headingCmd: steer.heading, speedCmd: speed)
    }

    /// Advance a monotone pointer to the path cell nearest the agent
    /// (bounded window), so validity checks only consider the part of
    /// the path still to be walked.
    static func remaining(_ cells: [Cell], _ ptr: Int, _ here: Cell) -> Int {
        var bestI = ptr
        var bestD: Int? = nil
        for i in ptr..<Swift.min(ptr + 12, cells.count) {
            let d = abs(Int(cells[i].x) - Int(here.x))
                + abs(Int(cells[i].y) - Int(here.y))
            if bestD == nil || d < bestD! {
                bestI = i
                bestD = d
            }
        }
        return bestI
    }

    // ---- mapping phase ---------------------------------------------------

    /// Walk the scan waypoints (optimistically planned) to build the
    /// map, like a user sweeping their phone around a room.
    @discardableResult
    public func runMapping(scanPts: [Vec], laps: Int = 2, perWp: Int = 350,
                           doneCheck: (() -> Bool)? = nil) -> Bool {
        try! fsm.step(.scanStarted)
        for _ in 0..<laps {
            for wp in scanPts {
                _ = gotoOptimistic(wp: wp, maxTicks: perWp)
            }
            sim.est.refreshClearance(force: true)
            if let doneCheck, doneCheck() {
                break
            }
        }
        try! fsm.step(.mapReady)
        return true
    }

    func gotoOptimistic(wp: Vec, maxTicks: Int) -> Bool {
        let est = sim.est
        var follower: GuidanceFollower? = nil
        var cells: [Cell]? = nil
        var ptr = 0
        var age = 1_000_000_000
        for _ in 0..<maxTicks {
            let changed = tickWorld()
            if dist(pos(), wp) < cfg.wpArrive {
                emit(changed)
                return true
            }
            age += 1
            let need = cells == nil || age >= cfg.mapReplanEvery
                || !pathValid(grid: est, cells: Array(cells![ptr...]),
                              params: optParams)
            var events: [String] = []
            var pathField: [Cell]? = nil
            var smoothedField: [Vec]? = nil
            if need {
                est.refreshClearance()
                let res = plan(grid: est, start: cell(),
                               goal: est.worldToCell(wp), params: optParams)
                replans += 1
                age = 0
                guard let res else {
                    emit(changed, events: ["WP_UNREACHABLE"])
                    return false  // try again next lap
                }
                cells = res.cells
                ptr = 0
                let smoothed = smooth(grid: est, cells: res.cells,
                                      params: optParams)
                follower = GuidanceFollower(
                    path: smoothed, lookahead: cfg.guidanceLookahead)
                events.append("PLANNED")
                pathField = res.cells
                smoothedField = smoothed
            }
            ptr = NavController.remaining(cells!, ptr, cell())
            let cue = follower!.cue(grid: est, pose: sim.pose)
            if cue.kind == .arrive {
                emit(changed, cue: cue, events: events)
                return true
            }
            let steer = steerMap.decide(
                grid: est, pose: sim.pose,
                goalBearing: bearing(from: pos(), to: cue.target))
            move(steer)
            if cue.kind == .offRoute {
                age = 1_000_000_000  // force replan next tick
            }
            emit(changed, cue: cue, steer: steer, events: events,
                 path: pathField, smoothed: smoothedField)
        }
        return false
    }

    // ---- explore phase (virtual-agent auto-mapping) ----------------------

    /// Frontier-driven auto-mapping: walk to the nearest frontier
    /// cluster until none remain.
    public func runExplore(maxTargets: Int = 60, perTarget: Int = 400,
                           minCluster: Int = 4) -> (targets: Int,
                                                    failures: Int) {
        let est = sim.est
        try! fsm.step(.scanStarted)
        let changed = tickWorld()          // open our eyes before asking
        est.refreshClearance(force: true)
        emit(changed, events: ["EXPLORE_STARTED"])
        var visited = 0
        var failures = 0
        while visited < maxTargets {
            est.refreshClearance(force: true)
            guard let tgt = selectTarget(grid: est, start: cell(),
                                         params: params,
                                         minCluster: minCluster,
                                         minDistM: cfg.wpArrive + 0.25)
            else {
                break
            }
            visited += 1
            if gotoOptimistic(wp: est.cellCenter(tgt.cell),
                              maxTicks: perTarget) {
                failures = 0
            } else {
                failures += 1
                if failures >= 3 {
                    break  // remaining clusters keep defeating us; stop
                }
            }
        }
        try! fsm.step(.mapReady)
        return (visited, failures)
    }

    // ---- walk mode (free roam, no destination) ---------------------------

    /// Walk mode: no destination — VFH picks the most open sector,
    /// hysteresis keeps the choice stable, and two safety gates guard
    /// the advance (see the Python docstrings for the measured
    /// rationale: freshness + swept-body vs live cell states).
    public func runWalk(ticks: Int) -> (traveledM: Double,
                                        blockedTicks: Int) {
        try! fsm.step(.walkToggled)
        var traveled = 0.0
        var blockedTicks = 0
        for _ in 0..<ticks {
            let changed = tickWorld()
            let before = pos()
            var steer = steerMap.decide(grid: sim.est, pose: sim.pose,
                                        goalBearing: nil)
            if !steer.blocked && (!frontFresh(heading: steer.heading)
                                  || !sweptClear(heading: steer.heading)) {
                steer = SteeringDecision(heading: steer.heading, speed: 0.0,
                                         freeDist: steer.freeDist,
                                         blocked: steer.blocked,
                                         reason: "hold_safe")
            }
            move(steer)
            traveled += dist(before, pos())
            blockedTicks += steer.blocked ? 1 : 0
            emit(changed, steer: steer)
        }
        try! fsm.step(.stop)
        return ((traveled * 100).rounded(.toNearestOrEven) / 100,
                blockedTicks)
    }

    /// Swept-body check against LIVE cell states. Gate radius sits just
    /// UNDER body radius on purpose (measured — see Python docstring).
    func sweptClear(heading: Double, advanceM: Double = 0.3,
                    pad: Double = -0.02) -> Bool {
        let est = sim.est
        let (x, y, _) = sim.pose
        let r = params.radius + pad
        let span = Int(r / est.cellSize) + 1
        let ca = cos(heading)
        let sa = sin(heading)
        let step = est.cellSize
        var d = step
        while d <= advanceM {
            let px = x + ca * d
            let py = y + sa * d
            let pc = est.worldToCell(Vec(px, py))
            for dy in -span...span {
                for dx in -span...span {
                    let c = Cell(pc.x + Int32(dx), pc.y + Int32(dy))
                    if est.inBounds(c) && est.state(c) == OCCUPIED
                        && dist(est.cellCenter(c), Vec(px, py)) <= r {
                        return false
                    }
                }
            }
            d += step
        }
        return true
    }

    /// True if every cell within distM along `heading` was observed
    /// within the last maxAge ticks.
    func frontFresh(heading: Double, distM: Double = 0.5,
                    maxAge: Int = 8) -> Bool {
        let est = sim.est
        let (x, y, _) = sim.pose
        let ca = cos(heading)
        let sa = sin(heading)
        let steps = max(1, Int(distM / est.cellSize))
        for i in 1...steps {
            let d = Double(i) * est.cellSize
            let c = est.worldToCell(Vec(x + ca * d, y + sa * d))
            if !est.inBounds(c) {
                return true          // the boundary is a wall, not a mover
            }
            if sim.tick - est.lastSeen[est.idx(c)] > maxAge {
                return false
            }
        }
        return true
    }

    // ---- guidance phase ("Ariadne mode") ---------------------------------

    /// Guide toward `goal` with incremental replanning. Returns true on
    /// arrival.
    public func runGuidance(goal: Vec, maxTicks: Int = 2500) -> Bool {
        let est = sim.est
        est.refreshClearance(force: true)
        let goalCell = est.worldToCell(goal)
        try! fsm.step(.goalSet)

        let ds = DStarLite(grid: est, start: cell(), goal: goalCell,
                           params: params)
        var follower: GuidanceFollower? = nil
        var cells: [Cell]? = nil
        var ptr = 0
        var boxed = 0
        var wait = 0

        func replan(_ events: inout [String]) -> [Cell]? {
            est.refreshClearance()
            let res = ds.plan()
            replans += 1
            guard let res else {
                if fsm.state != .blocked {
                    try! fsm.step(.planFailed)
                }
                events.append("PLAN_FAILED")
                return nil
            }
            let smoothed = smooth(grid: est, cells: res.cells,
                                  params: params)
            follower = GuidanceFollower(path: smoothed,
                                        lookahead: cfg.guidanceLookahead)
            ptr = 0
            try! fsm.step(.planReady)
            events.append("REROUTED")
            return res.cells
        }

        var events: [String] = []
        cells = replan(&events)
        emit([], events: events, path: cells, smoothed: follower?.path)

        for _ in 0..<maxTicks {
            let changed = tickWorld()
            if !changed.isEmpty {
                ds.notifyChanged(changed)
            }
            let here = cell()
            if here != ds.start {
                ds.updateStart(here)
            }

            events = []
            var pathField: [Cell]? = nil
            var smoothedField: [Vec]? = nil

            if fsm.state == .blocked {
                wait += 1
                if wait >= cfg.blockedWait {
                    wait = 0
                    try! fsm.step(.retry)   // BLOCKED -> PLANNING
                    cells = replan(&events)
                    pathField = cells
                    smoothedField = cells != nil ? follower?.path : nil
                }
                emit(changed, events: events, path: pathField,
                     smoothed: smoothedField)
                continue
            }

            ptr = NavController.remaining(cells!, ptr, here)
            if !pathValid(grid: est, cells: Array(cells![ptr...]),
                          params: params) {
                try! fsm.step(.routeBlocked)   // GUIDING -> PLANNING
                cells = replan(&events)
                pathField = cells
                smoothedField = cells != nil ? follower?.path : nil
                if cells == nil {
                    emit(changed, events: events)
                    continue
                }
            }

            var cue = follower!.cue(grid: est, pose: sim.pose)
            if cue.kind == .arrive {
                try! fsm.step(.arrivedEvt)
                emit(changed, cue: cue, events: events + ["ARRIVED"],
                     path: pathField, smoothed: smoothedField)
                return true
            }
            if cue.kind == .offRoute {
                try! fsm.step(.offRoute)      // GUIDING -> PLANNING
                cells = replan(&events)
                pathField = cells
                smoothedField = cells != nil ? follower?.path : nil
                if cells == nil {
                    emit(changed, events: events)
                    continue
                }
                cue = follower!.cue(grid: est, pose: sim.pose)
            }

            let steer = steerGuide.decide(
                grid: est, pose: sim.pose,
                goalBearing: bearing(from: pos(), to: cue.target))
            if steer.blocked {
                boxed += 1
                if boxed >= cfg.blockedSteerLimit {
                    boxed = 0
                    try! fsm.step(.routeBlocked)
                    cells = replan(&events)
                    pathField = cells
                    smoothedField = cells != nil ? follower?.path : nil
                }
            } else {
                boxed = 0
            }
            move(steer)
            emit(changed, cue: cue, steer: steer, events: events,
                 path: pathField, smoothed: smoothedField)
        }
        return false
    }
}
